internal/cli/kraft/pkg/packager_kraftfile_runtime.go

```diff
@@ -110,95 +113,181 @@ func (p *packagerKraftfileRuntime) Pack(ctx context.Context, opts *PkgOptions, a
+		// catalog with the host architecture and platform.
 
-					qopts = append(qopts,
-						packmanager.WithPlatform(opts.Platform),
-						packmanager.WithArchitecture(opts.Architecture),
-					)
+		if opts.Architecture == "" {
+			opts.Architecture, err = ukarch.HostArchitecture()
+			if err != nil {
+				return nil, fmt.Errorf("could not get host architecture: %w", err)
+			}
+		}
+		if opts.Platform == "" {
+			plat, _, err := platform.Detect(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("could not get host platform: %w", err)
+			}
+
+			opts.Platform = plat.String()
+		}
+
+		qopts = append(qopts,
+			packmanager.WithPlatform(opts.Platform),
+			packmanager.WithArchitecture(opts.Architecture),
+		)
 
+		treemodel, err := processtree.NewProcessTree(
+			ctx,
+			[]processtree.ProcessTreeOption{
+				processtree.IsParallel(false),
+				processtree.WithRenderer(
+					log.LoggerTypeFromString(config.G[config.KraftKit](ctx).Log.Type) != log.FANCY,
+				),
+				processtree.WithFailFast(true),
+				processtree.WithHideOnSuccess(true),
+			},
+			processtree.NewProcessTreeItem(
+				fmt.Sprintf(
+					"searching for %s:%s",
+					opts.Project.Runtime().Name(),
+					opts.Project.Runtime().Version(),
+				),
+				"",
+				func(ctx context.Context) error {
 					packs, err = opts.pm.Catalog(ctx, qopts...)
 					if err != nil {
 						return fmt.Errorf("could not query catalog: %w", err)
-					} else if len(packs) == 0 {
-						return fmt.Errorf("coud not find runtime '%s'", opts.Project.Runtime().Name())
-					} else if len(packs) > 1 {
-						return fmt.Errorf("could not find runtime: too many options")
 					}
+					return nil
+				},
+			),
+		)
+		if err != nil {
+			return nil, err
+		}
 
-					log.G(ctx).
-						WithField("arch", opts.Architecture).
-						WithField("plat", opts.Platform).
-						Info("using")
-				}
+		if err := treemodel.Start(); err != nil {
+			return nil, err
+		}
+	}
 
-				return nil
-			},
-		),
-	)
-	if err != nil {
-		return nil, err
+	if len(packs) == 0 {
+		return nil, fmt.Errorf(
+			"coud not find runtime '%s:%s' (%s/%s)",
+			opts.Project.Runtime().Name(),
+			opts.Project.Runtime().Version(),
+			opts.Platform,
+			opts.Architecture,
+		)
+	} else if len(packs) == 1 {
+		selected = &packs[0]
+	} else if len(packs) > 1 {
+		selected, err = selection.Select[pack.Package]("multiple runtimes available", packs...)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	if err := treemodel.Start(); err != nil {
-		return nil, err
+	log.G(ctx).
+		WithField("arch", opts.Architecture).
+		WithField("plat", opts.Platform).
+		Info("using")
+
+	runtime := *selected
+	pulled, _, _ := runtime.PulledAt(ctx)
+
+	// Temporarily save the runtime package.
+	if err := runtime.Save(ctx); err != nil {
+		return nil, fmt.Errorf("could not save runtime package: %w", err)
+	}
+
+	// Remove the cached runtime package reference if it was not previously
+	// pulled.
+	if !pulled {
+		defer runtime.Delete(ctx) //nolint:errcheck
```

Comment:
This is specifically hidden because it is a fail-safe, catch-all.  I can however add it to the trace logs to keep it available, but it will almost always fire.